### PR TITLE
feat: migration from mswjs to internal solution

### PR DESCRIPTION
**Pull request by kurpav**

### CODE REVIEW

File: src/interceptor/NodeClientRequest.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            status: response.statusCode || 200,
+            statusText: response.statusMessage || 'OK',
+            headers: createHeadersFromIncomingHttpHeaders(message.headers),
+            body: responseBody
+          },
+          requestId
+        );
+      }
+
+      if (!this.ignoredDomains.includes(this.url.hostname)) {
```
Comment (aklarfeld) [inline src/interceptor/NodeClientRequest.ts:123]: can you hoist this check into a helper function that can also accept an option to ignore internal domains?

Comment (aklarfeld) [reply src/interceptor/NodeClientRequest.ts:123]: ignoring internal domains should be default 'true' and I realize the check is imperfect, but I was hoping we can do something like this:

```export function isLocalHostname (hostname: string): boolean {
  if(!hostname) return true;
  const [domain, tld] = hostname.split('.')
  if(!tld) {
    return true
  }
  if(['local'].includes(tld)) {
    return true
  }
  return false;
}
```

File: src/interceptor/utils/cloneObject.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  logger.debug('checking the object constructor:', obj.constructor.name);
+  return obj.constructor.name === 'Object';
+}
+
+export function cloneObject<ObjectType extends Record<string, any>>(
+  obj: ObjectType
+): ObjectType {
+  logger.debug('cloning object:', obj);
```
Comment (aklarfeld) [inline src/interceptor/utils/cloneObject.ts:20]: can you check and/or remove these debugging statements if the payloads are getting stored into memory regardless of whether debug is turned on or not?

**Review by aklarfeld — CHANGES_REQUESTED** [review]
This looks pretty sweet! Thank you for handling it. Two additions:

- Can we support native `fetch` for node 18?
- Can you hoist the ignoredDomains check into a higher level function so we can add an option to ignore local domains?

I will deploy this to our staging environment and come back with any issues we might find.

**Comment by aklarfeld** [comment]
I made the hoisting changes and ignore local domains by default.

File: src/interceptor/utils/cloneObject.ts (hunk 2 above)
Comment (kurpav) [reply src/interceptor/utils/cloneObject.ts:20]: I disabled it for now. I left the logs to allow local debugging, so they shouldn't affect production builds

**Pull request merged**